Fix support for CPUS with Always Running Timers (ART)

mrmiller opened the pull request:
OcCpuLib: Previously, we assumed all CPUs had the same core crystal clock frequency for their Always Running Timer (ART), if present: 24 Mhz. Xeon Scalable and Intel Atoms with Goldmont architecture have different frequencies (25 Mhz and 19.2 Mhz respectively). If the CPU supports reporting its frequency, use that. Otherwise, fallback to a value based on the CPU's model and family.

Cleaned up the logic for how we compute our CPU and FSB frequencies. For CPUs that support ART, use that instead of the TSC frequency.

Added documentation and renamed `OC_CPU_INFO::TSCFrequency` to `CPUFrequencyFromTSC` and added `CPUFrequencyFromART` for clarity. These are intermediate values used to compute `CPUFrequency`. `ARTFrequency` is now correctly set to the core crystal clock frequency rather than `CPUFrequencyFromART`, which it was previously.

Lastly, adjusted the `CPUFrequencyFromART` calculation to include the TSC offset in `MSR_IA32_TSC_ADJUST` if present. I'm not aware of which CPUs have a non-zero offset but it's part of Intel's calculation for computing the CPU frequency based on the ART.

vit9696 commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -956,8 +982,8 @@ OcCpuScanProcessor (
     "OCCPU: %a %a %11lld %5dMHz\n",
```
There is an issue with formatting again.

mrmiller replied:
Fixed in latest.

vit9696 commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -661,46 +663,70 @@ ScanIntelProcessor (
       DEBUG ((
         DEBUG_INFO,
-        "OCCPU: %a %a %11lld %5dMHz %u * %u / %u = %ld\n",
+        "OCCPU: %a %a %11llu %5dMHz = %u * %u / %u + %u\n",
```
Again issues with types, should not it be:
`"OCCPU: %a %a %11Lu %5LuMHz = %Lu * %u / %u + %Lu\n",`

mrmiller replied:
This should be fixed.

vit9696 commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -661,46 +663,70 @@ ScanIntelProcessor (
+      DEBUG ((DEBUG_INFO, "OCCPU: TSC Adjust %llu\n", TscAdjust ));
+      ASSERT (Cpu->ARTFrequency > 0ULL);
+      Cpu->CPUFrequencyFromART = MultU64x32 (Cpu->ARTFrequency, (UINT32) DivU64x32 (CpuidEbx, CpuidEax)) + TscAdjust;
 
```
This looks wrong to me.

Firstly because the calculation may overflow, and thus I want it to rely on `MultThenDivU64x64x32` I posted [here](https://github.com/acidanthera/bugtracker/issues/448#issuecomment-519468023). Would you please add it to `OcMiscLib`?

Secondly because `TscAdjust` is in TSC values. I reread the docs, and it is just an offset to get a TSC value from ART. CPU frequency will not be affected with it, as these are two different calculations:

```
CPU frequency = ART frequency * Ebx / Eax
TSC timestamp = ART timestamp * Ebx / Eax + TSC offset
```

ART timestamps are provided by some onboard hardware. I believe the formula in `CPUFrequencyFromART` description needs to be changed for this reason as well.

mrmiller replied:
Good catch! I added `MultThenDivU64x64x32` to `OcMiscLib` in a new file called `Math.c`. I adjusted the ART-based CPU frequency calculation and adjusted the documentation to match.

vit9696 commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -661,46 +663,70 @@ ScanIntelProcessor (
+        //
+        // TODO: Xeon Ws share the same CPUID but I believe they have 24 Mhz ART
+        //
+        Cpu->ARTFrequency = 25000000ULL; // 25 Mhz
```
Yes, I can confirm that Xeon W, Xeon X, and Xeon Scalable have all the same Platform ID and CPUID but different frequency. This is terrible, and I am not sure how we can handle it. As a last resort option we can use CPUID brand string, but I hope that something better arises. I will check the docs when I get a spare minute.

mrmiller replied:
For now, I adjusted `DetectAppleMajorType` to return a new `enum` value `AppleProcessorMajorXeonScalable` (determined from the brand string) which we can use to differentiate. I believe that `enum` is for internal use only anyways so that should be fine. I pun the `AppleProcessorMajorXeonScalable` into a Xeon W in `DetectAppleProcessorType` now.

I don't think it makes the code much uglier, thankfully.

vit9696 replied:
Hmm, issue is, `AppleProcessorMajorXeonW` is not internal, it is Apple value actually, and they do not have Xeon Scalable. Thus the reason it is in a separate EfiPkg repository. If anything, we should at least name it differently.

I need to really think if there is any better way to distinguish the CPUs. Currently I dislike the fact to need to check the brand string, though it will work, yes.

mrmiller replied:
Alright, I rolled that back. I've tacked it onto `DetectAppleMajorType` as a `BOOLEAN` additional `OUT` argument called `IsXeonScalable` until we can come up with a more permanent or better solution.

mrmiller commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -661,46 +663,70 @@ ScanIntelProcessor (
     if (CpuidEax > 0 && CpuidEbx > 0) {
-      Cpu->CPUFrequency = MultU64x32 (BASE_ART_CLOCK_SOURCE, (UINT32) DivU64x32 (CpuidEbx, CpuidEax));
+      TscAdjust = AsmReadMsr64 (MSR_IA32_TSC_ADJUST);
+      DEBUG ((DEBUG_INFO, "OCCPU: TSC Adjust %llu\n", TscAdjust ));
```
Fixed in latest. I did notice other format strings where the variable width didn't exactly match the specifier. The docs say that `%u`/`%d` implies `int` yet there are lots of places passing `UINT16` (`unsigned short`) or even `BOOLEAN` (`unsigned char`) sometimes. Is that okay or should we be casting them to `UINT32`?

vit9696 commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -953,16 +981,16 @@ OcCpuScanProcessor (
   DEBUG ((
     DEBUG_INFO,
-    "OCCPU: %a %a %11lld %5dMHz\n",
+    "OCCPU: %a %a %11,LuHz %5LuMHz\n",
```
`%11,Lu` is that from that printf? Honestly, I do not see how it is supposed to work…

mrmiller replied:
It's not standard for `printf` but it is part of the [EDK2 format string spec for Print Lib](https://github.com/tianocore/edk2/blob/3806e1fd139775610d8f2e7541a916c3a91ad989/MdePkg/Include/Library/PrintLib.h#L39) you sent me:
```
- ,
  - Place a comma every 3rd digit of the number.  Only valid for type d.
    If 0 is also specified, then 0 is ignored.
```

I tested the output and it prints `3,000,000,000` instead of `3000000000`. Now, whether the `11` applies before or after the commas are added, you've got me. That I didn't check.

mrmiller replied:
Though reading this, perhaps it's supposed to be `%,14Lu` (to account for the commas) instead. Will change.

vit9696 replied:
They say it is only valid for d, and not u, but I guess they lie. To be honest, in Russia and Germany comma works as a floating point separator, and this change may make it quite confusing in both log and code reading.

If you think it is still worth trying to preserve it, could probably check how edk2 guys do it. I am away from keyboard at the moment, and github does not let me search commas.

mrmiller replied:
Eh, I'll just remove it. It's only a minor convenience to break up the characters. The docs don't say any of the flags work with `u`, only `d`. `u` goes completely unmentioned, including when discussing the `L` flag:
```
- L, l
  - The number being printed is size UINT64.  Only valid for types X, x, and d.
    If this flag is not specified, then the number being printed is size int.
```
I assume it's just a mistake in the docs but maybe it should be changed to `%11Ld` even.

vit9696 replied:
Just a mistake in the docs in my opinion. Should report that sooner or later I guess. Go with `%11Lu`.

mrmiller replied:
Fixed.

vit9696 commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -1095,7 +1123,7 @@ OcIsSandyOrIvy (
   DEBUG ((
     DEBUG_VERBOSE,
-    "OCCPU: Discovered CpuFamily %d CpuModel %d SandyOrIvy %d\n",
+    "OCCPU: Discovered CpuFamily %d CpuModel %d SandyOrIvy %a\n",
```
SandyOrIvy is boolean, thus decimal, and not an ASCII string.

mrmiller replied:
Doh! I had started going through and fixing up all the format strings and rolled some of them back but forgot this one. I was going to change it to an ASCII string and then pass either `SandyOrIvy ? "YES" : "NO"`. This is one of those places I mentioned in a previous comment where the code is passing an `unsigned char` (`BOOLEAN`) to the formatter `%d`. I'm not sure that's okay as it certainly wouldn't be okay if this were `printf`.

vit9696 replied:
Mhm, why is that not ok with C printf? Will not integer promotion promote it to int, thus %d? I think we use %d in many places for BOOLEn, and it makes sense to me to continue to to be honest.

mrmiller replied:
Huh, today I learned. I didn't realize types smaller than `int` were promoted to `int` in variadic function arguments and functions without prototypes. Ignore everything I was saying about the use of `UINT16` and `BOOLEAN` with `%d`. I'm an idiot.

vit9696 commented on Library/OcCpuLib/OcCpuLib.c:
```diff
@@ -661,46 +664,71 @@ ScanIntelProcessor (
+      DEBUG ((DEBUG_INFO, "OCCPU: TSC Adjust %Lu\n", TscAdjust ));
+
+      ASSERT (Cpu->ARTFrequency > 0ULL);
+      Cpu->CPUFrequencyFromART = MultThenDivU64x64x32 (Cpu->ARTFrequency, CpuidEbx, (UINT32) CpuidEax, NULL);
```
`CpuidEax` is already `(UINT32)`, so no need to cast. There also is an extra space in `TscAdjust ))` part.

mrmiller replied:
Fixed in latest. Should I be resolving these or you? (I assume the latter but just double-checking.)

vit9696 commented on Library/OcMiscLib/Math.c:
```diff
@@ -0,0 +1,60 @@
+ **/
+UINT64
+MultThenDivU64x64x32 (
+                      IN      UINT64                    Multiplicand,
```
I get a feeling something got borked with the spacing here, could make it more usual?
Two spaces till IN/OUT, then up to 1 to type, and up to 2 to arguments… Like

```C
VOID
DataHubLogData (
  IN  EFI_GUID  *DataRecordGuid,
  IN  EFI_GUID  *ProducerName,
  IN  VOID      *RawData,
  IN  UINT32    RawDataSize
  );
```

mrmiller replied:
Fixed in latest.

vit9696 commented:
I am currently checking how Linux fights with the clock issue, and it seems that they use leaf 16, despite the fact that Intel recommends it for information only purposes. I wonder whether it actually is a better idea than to hardcode based on brand. Do not want to hurry with this, needs proper thought and decision.

See: https://github.com/torvalds/linux/blob/223cea6a4f0552b86fb25e3b8bbd00469816cd7a/arch/x86/kernel/tsc.c#L652-L663

Are these values available in your CPU?

mrmiller commented:
> I am currently checking how Linux fights with the clock issue, and it seems that they use leaf 16, despite the fact that Intel recommends it for information only purposes. I wonder whether it actually is a better idea than to hardcode based on brand. Do not want to hurry with this, needs proper thought and decision.

Well, the brand special case is a fallback. There is support there for querying the core crystal clock frequency if the CPU supports it. Only if that mechanism isn't available do we fall back to identifying the model and family, and then only the Skylake chips do we differentiate by brandstring. In this case, we're technically doing exactly what Intel is suggesting and what you initially sent me in `18.7.3 Determining the Processor Base Frequency`.

> See: https://github.com/torvalds/linux/blob/223cea6a4f0552b86fb25e3b8bbd00469816cd7a/arch/x86/kernel/tsc.c#L652-L663
> 
> Are these values available in your CPU?

I will check.

vit9696 commented:
I checked other operating systems too, and the situation is as follows:

- FreeBSD has some junk with parsing brand string and only uses 15h leaf values when [available](https://github.com/freebsd/freebsd/blob/250e158/sys/x86/x86/tsc.c#L142).
- NetBSD does not care at all and uses delays.
- OpenBSD is doing [predefined frequency](https://github.com/openbsd/src/blob/8d9f4b3/sys/arch/amd64/amd64/tsc.c#L57) when unknown but not for 06_55h CPUID

Linux actually had two patches for the above code.

Current:
https://github.com/torvalds/linux/commit/604dc9170f2435d27da5039a3efd757dceadc684#diff-96989d5a06fa45e5014d1613838a4f86
https://lore.kernel.org/lkml/20190509055417.13152-1-drake@endlessm.com/
https://lore.kernel.org/lkml/20190419083533.32388-1-drake@endlessm.com/ 

Older:
https://github.com/torvalds/linux/commit/b511203093489eb1829cb4de86e8214752205ac6#diff-96989d5a06fa45e5014d1613838a4f86
https://lkml.kernel.org/r/ff6dcea166e8ff8f2f6a03c17beab2cb436aa779.1513920414.git.len.brown@intel.com

Given the description and the whole situation with Intel mess, I suggest us to take the following approach to determine `ARTFrequency`:
1. Replace >= Skylake model checking by CPUID max level comparison against `CPUID_TIME_STAMP_COUNTER`. Intel does not guarantee to increment new models, they are chaotic. Leave as 0 otherwise.
1. Try obtaining crystal clock frequency from `CPUID_TIME_STAMP_COUNTER` (15h) leaf from ECX.
1. Take [predefined values](https://github.com/openbsd/src/blob/8d9f4b3/sys/arch/amd64/amd64/tsc.c#L67) from OpenBSD, including 24 MHz ones if failed.
1. Check for `CPUID_PROCESSOR_FREQUENCY` availability and use it [like Linux](https://github.com/torvalds/linux/commit/604dc9170f2435d27da5039a3efd757dceadc684#diff-96989d5a06fa45e5014d1613838a4f86R660).
1. Assume 24 MHz otherwise.

Now we can calculate CPUFrequencyFromART as usual. 4th step makes us basically have CPUFrequencyFromART = `CPUID_PROCESSOR_FREQUENCY.EAX` * 100, but that can live, as we need ARTFrequency independently to report to macOS anyway.

This should cover all the cases including yours, but just in case I would add a DEBUG_WARN print when CPUFrequencyFromART is not 0 and differs from CPUFrequencyFromTSC by more than 50 MHz.

This is just a draft, and I would personally want your opinion, and any others that may jump in first. CC @Download-Fritz

mrmiller commented:
That all sounds reasonable to me. I’ll look into modifying it as suggested.

On my end, I found the docs talking about the Xeon Scalable family and CPU
identification. I couldn’t find anything for Xeon W though, so if you
happen to come across the Intel reference for that, let me know.
-- 
http://mrmiller.net Sent from my phone

vit9696 commented:
@mrmiller I talked with @Download-Fritz, and we think that the suggested approach should do if it works.

Normally CPU identification is published in a specification update for the CPU, so you can find it with `inurl:https://www.intel.com/content/dam/www/public/us/en/documents/specification-update` or `site:intel.com Component Identification via Programming Interface`, but for Xeon W it is not the case. Perhaps Intel thinks that `xeon-scalable-spec-update.pdf` is common for both Scalable and Xeon W :x.

vit9696 commented:
Intel is currently contributing a new TimerLib based on ART to EDK II, and they explained why Xeon W models have 24 MHz crystal clock frequency and Xeon Scalable ones have 25 MHz. The point is vertical segment: Atoms get 19.2, client get 24, server get 25. They might be able to update SDM if this holds true for all relevant cases.

Although this makes sense, I am not positive whether it gives us any advantage in terms of improving the suggested algorithm. In any case, I asked them to explain how to programmatically identify the vertical segment, and there is a non-zero chance we get some answer.

mrmiller commented:
Alright, updated based on the above procedure. I added a new field for the maximum leaf and fixed two minor bugs with that (they were checking standard CPUID leaf numbers against the maximum extended leaf).

My Xeon Golds don't report the core crystal clock frequency via `CPUID_TIME_STAMP_COUNTER` `ECX`. OC gets 25 Mhz correctly using `CPUID_PROCESSOR_FREQUENCY` though.

vit9696 commented:
Ugh, the hell github is doing :x Sorry for that. I rebased your pull request but actually could only partially push it to your repository. Moved here: https://github.com/acidanthera/OcSupportPkg/pull/10. Will review and merge soon, thanks!

mrmiller commented:
Sorry about that! I'm not quite sure what was happening with GitHub and the rebasing. It all looked fine on my local copy but went haywire as soon as I pushed to the remote.
